File: frontend/src/api/student/library.ts
import { API_BASE_URL } from '../config';

export interface Resource {
  id: number;
  title: string;
  description: string;
  type: 'video' | 'document' | 'image' | 'audio' | 'link' | 'interactive';
  subject: string;
  level: string;
  tags: string[];
  author: string;
  created_at: string;
  duration?: number; // en minutes pour les vidéos
  file_size?: string;
  views: number;
  rating: number;
  is_favorite: boolean;
  is_in_collection: boolean;
  collections: string[];
  url: string;
  thumbnail?: string;
}

export interface Collection {
  id: number;
  name: string;
  description: string;
  resource_count: number;
  is_public: boolean;
  created_at: string;
  color: string;
}

export interface Subject {
  id: number;
  name: string;
  color: string;
  resource_count: number;
}

export interface Level {
  id: number;
  name: string;
  color: string;
  resource_count: number;
}

export interface CreateCollectionRequest {
  name: string;
  description: string;
  is_public: boolean;
  color?: string;
}

export interface UpdateCollectionRequest {
  name?: string;
  description?: string;
  is_public?: boolean;
  color?: string;
}

export interface AddToCollectionRequest {
  resource_id: number;
  collection_id: number;
}

export interface RateResourceRequest {
  resource_id: number;
  rating: number; // 1-5
  comment?: string;
}

class LibraryAPI {
  private baseURL = `${API_BASE_URL}/library`;

  // ===== RESSOURCES =====

  // Récupérer toutes les ressources
  async getResources(
    subject?: string,
    level?: string,
    type?: string,
    searchTerm?: string,
    sortBy?: 'recent' | 'popular' | 'rating' | 'title',
    sortOrder?: 'asc' | 'desc'
  ): Promise<Resource[]> {
    try {
      const params = new URLSearchParams();
      if (subject) params.append('subject', subject);
      if (level) params.append('level', level);
      if (type) params.append('type', type);
      if (searchTerm) params.append('search', searchTerm);
      if (sortBy) params.append('sort_by', sortBy);
      if (sortOrder) params.append('sort_order', sortOrder);

      const response = await fetch(`${this.baseURL}/resources?${params}`, {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${localStorage.getItem('token')}`
        }
      });

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      return await response.json();
    } catch (error) {
      console.error('Error fetching resources:', error);
      throw error;
    }
  }

  // Récupérer une ressource spécifique
  async getResource(resourceId: number): Promise<Resource> {
    try {
      const response = await fetch(`${this.baseURL}/resources/${resourceId}`, {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${localStorage.getItem('token')}`
        }
      });

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      return await response.json();
    } catch (error) {
      console.error('Error fetching resource:', error);
      throw error;
    }
  }

  // Marquer une ressource comme vue
  async markAsViewed(resourceId: number): Promise<void> {
    try {
      const response = await fetch(`${this.baseURL}/resources/${resourceId}/view`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${localStorage.getItem('token')}`
        }
      });

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }
    } catch (error) {
      console.error('Error marking resource as viewed:', error);
      throw error;
    }
  }

  // ===== FAVORIS =====

  // Récupérer les ressources favorites
  async getFavoriteResources(): Promise<Resource[]> {
    try {
      const response = await fetch(`${this.baseURL}/favorites`, {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${localStorage.getItem('token')}`
        }
      });

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      return await response.json();
    } catch (error) {
      console.error('Error fetching favorite resources:', error);
      throw error;
    }
  }

  // Marquer/démarquer une ressource comme favorite
  async toggleFavorite(resourceId: number): Promise<void> {
    try {
      const response = await fetch(`${this.baseURL}/resources/${resourceId}/favorite`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${localStorage.getItem('token')}`
        }
      });

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }
    } catch (error) {
      console.error('Error toggling resource favorite:', error);
      throw error;
    }
  }

  // Ajouter/retirer une ressource d'une collection
  async toggleCollection(resourceId: number): Promise<void> {
    try {
      const response = await fetch(`${this.baseURL}/resources/${resourceId}/collection`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${localStorage.getItem('token')}`
        }
      });

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }
    } catch (error) {
      console.error('Error toggling resource collection:', error);
      throw error;
    }
  }

  // ===== COLLECTIONS =====

  // Récupérer toutes les collections
  async getCollections(): Promise<Collection[]> {
    try {
      const response = await fetch(`${this.baseURL}/collections`, {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${localStorage.getItem('token')}`
        }
      });

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      return await response.json();
    } catch (error) {
      console.error('Error fetching collections:', error);
      throw error;
    }
  }

  // Créer une nouvelle collection
  async createCollection(data: CreateCollectionRequest): Promise<Collection> {
    try {
      const response = await fetch(`${this.baseURL}/collections`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${localStorage.getItem('token')}`
        },
        body: JSON.stringify(data)
      });

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      return await response.json();
    } catch (error) {
      console.error('Error creating collection:', error);
      throw error;
    }
  }

  // Mettre à jour une collection
  async updateCollection(collectionId: number, data: UpdateCollectionRequest): Promise<Collection> {
    try {
      const response = await fetch(`${this.baseURL}/collections/${collectionId}`, {
        method: 'PUT',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${localStorage.getItem('token')}`
        },
        body: JSON.stringify(data)
      });

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      return await response.json();
    } catch (error) {
      console.error('Error updating collection:', error);
      throw error;
    }
  }

  // Supprimer une collection
  async deleteCollection(collectionId: number): Promise<void> {
    try {
      const response = await fetch(`${this.baseURL}/collections/${collectionId}`, {
        method: 'DELETE',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${localStorage.getItem('token')}`
        }
      });

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }
    } catch (error) {
      console.error('Error deleting collection:', error);
      throw error;
    }
  }

  // Récupérer les ressources d'une collection
  async getCollectionResources(collectionId: number): Promise<Resource[]> {
    try {
      const response = await fetch(`${this.baseURL}/collections/${collectionId}/resources`, {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${localStorage.getItem('token')}`
        }
      });

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      return await response.json();
    } catch (error) {
      console.error('Error fetching collection resources:', error);
      throw error;
    }
  }

  // Ajouter une ressource à une collection
  async addToCollection(data: AddToCollectionRequest): Promise<void> {
    try {
      const response = await fetch(`${this.baseURL}/collections/${data.collection_id}/resources`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${localStorage.getItem('token')}`
        },
        body: JSON.stringify({ resource_id: data.resource_id })
      });

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }
    } catch (error) {
      console.error('Error adding resource to collection:', error);
      throw error;
    }
  }

  // Retirer une ressource d'une collection
  async removeFromCollection(collectionId: number, resourceId: number): Promise<void> {
    try {
      const response = await fetch(`${this.baseURL}/collections/${collectionId}/resources/${resourceId}`, {
        method: 'DELETE',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${localStorage.getItem('token')}`
        }
      });

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }
    } catch (error) {
      console.error('Error removing resource from collection:', error);
      throw error;
    }
  }

  // ===== MATIÈRES ET NIVEAUX =====

  // Récupérer toutes les matières
  async getSubjects(): Promise<Subject[]> {
    try {
      const response = await fetch(`${this.baseURL}/subjects`, {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${localStorage.getItem('token')}`
        }
      });

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      return await response.json();
    } catch (error) {
      console.error('Error fetching subjects:', error);
      throw error;
    }
  }

  // Récupérer tous les niveaux
  async getLevels(): Promise<Level[]> {
    try {
      const response = await fetch(`${this.baseURL}/levels`, {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${localStorage.getItem('token')}`
        }
      });

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      return await response.json();
    } catch (error) {
      console.error('Error fetching levels:', error);
      throw error;
    }
  }

  // ===== ÉVALUATIONS =====

  // Évaluer une ressource
  async rateResource(data: RateResourceRequest): Promise<void> {
    try {
      const response = await fetch(`${this.baseURL}/resources/${data.resource_id}/rate`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${localStorage.getItem('token')}`
        },
        body: JSON.stringify({
          rating: data.rating,
          comment: data.comment
        })
      });

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }
    } catch (error) {
      console.error('Error rating resource:', error);
      throw error;
    }
  }

  // Récupérer les évaluations d'une ressource
  async getResourceRatings(resourceId: number): Promise<{
    average_rating: number;
    total_ratings: number;
    ratings: Array<{
      user_id: number;
      rating: number;
      comment?: string;
      created_at: string;
    }>;
  }> {
    try {
      const response = await fetch(`${this.baseURL}/resources/${resourceId}/ratings`, {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${localStorage.getItem('token')}`
        }
      });

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      return await response.json();
    } catch (error) {
      console.error('Error fetching resource ratings:', error);
      throw error;
    }
  }

  // ===== RECHERCHE ET FILTRES =====

  // Recherche avancée
  async searchResources(query: string, filters?: {
    subject?: string;
    level?: string;
    type?: string;
    min_rating?: number;
    max_duration?: number;
  }): Promise<Resource[]> {
    try {
      const params = new URLSearchParams();
      params.append('query', query);
      
      if (filters) {
        Object.entries(filters).forEach(([key, value]) => {
          if (value !== undefined) {
            params.append(key, value.toString());
          }
        });
      }

      const response = await fetch(`${this.baseURL}/search?${params}`, {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${localStorage.getItem('token')}`
        }
      });

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      return await response.json();
    } catch (error) {
      console.error('Error searching resources:', error);
      throw error;
    }
  }

  // Récupérer les ressources recommandées
  async getRecommendedResources(limit: number = 10): Promise<Resource[]> {
    try {
      const response = await fetch(`${this.baseURL}/recommendations?limit=${limit}`, {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${localStorage.getItem('token')}`
        }
      });

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      return await response.json();
    } catch (error) {
      console.error('Error fetching recommended resources:', error);
      throw error;
    }
  }

  // ===== STATISTIQUES =====

  // Récupérer les statistiques de la bibliothèque
  async getLibraryStats(): Promise<{
    total_resources: number;
    total_views: number;
    total_favorites: number;
    total_collections: number;
    most_viewed_subject: string;
    most_rated_resource: string;
  }> {
    try {
      const response = await fetch(`${this.baseURL}/stats`, {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${localStorage.getItem('token')}`
        }
      });

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      return await response.json();
    } catch (error) {
      console.error('Error fetching library stats:', error);
      throw error;
    }
  }

  // Récupérer l'historique de consultation
  async getViewHistory(limit: number = 20): Promise<Array<{
    resource: Resource;
    viewed_at: string;
  }>> {
    try {
      const response = await fetch(`${this.baseURL}/history?limit=${limit}`, {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${localStorage.getItem('token')}`
        }
      });

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      return await response.json();
    } catch (error) {
      console.error('Error fetching view history:', error);
      throw error;
    }
  }
}

export const libraryAPI = new LibraryAPI(); 

File: backend/api/v1/library.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from typing import List, Optional
from datetime import datetime
import json

from core.database import SessionLocal
from api.v1.users import get_current_user
from models.user import User
from models.library import UserFavorite, Collection, CollectionItem, ContentHistory, ContentRecommendation, Playlist, PlaylistItem
from models.content import Content
from models.class_group import ClassGroup

def get_db():
    db = SessionLocal()
    try:
        yield db
    finally:
        db.close()

router = APIRouter(prefix="/library", tags=["Library"])

# =====================================================
# ENDPOINTS POUR LES RESSOURCES
# =====================================================

@router.get("/resources", response_model=List[dict])
def get_resources(
    subject: Optional[str] = None,
    level: Optional[str] = None,
    type: Optional[str] = None,
    search: Optional[str] = None,
    limit: int = 50,
    offset: int = 0,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Récupérer les ressources de la bibliothèque avec filtres"""
    query = db.query(Content)
    
    if subject:
        query = query.filter(Content.subject == subject)
    if level:
        query = query.filter(Content.level == level)
    if type:
        query = query.filter(Content.type == type)
    if search:
        query = query.filter(
            (Content.title.contains(search)) | 
            (Content.description.contains(search))
        )
    
    resources = query.offset(offset).limit(limit).all()
    
    # Convertir en format compatible avec le frontend
    result = []
    for resource in resources:
        # Vérifier si l'utilisateur a mis en favori
        is_favorite = db.query(UserFavorite).filter(
            UserFavorite.user_id == current_user.id,
            UserFavorite.content_id == resource.id
        ).first() is not None
        
        resource_data = {
            "id": resource.id,
            "title": resource.title,
            "description": resource.description,
            "type": resource.type,
            "subject": resource.subject,
            "level": resource.level,
            "tags": json.loads(resource.tags) if resource.tags else [],
            "author": resource.author,
            "created_at": resource.created_at.isoformat() if resource.created_at else "",
            "duration": getattr(resource, 'duration', None),
            "file_size": getattr(resource, 'file_size', None),
            "views": getattr(resource, 'views', 0),
            "rating": getattr(resource, 'rating', 0.0),
            "is_favorite": is_favorite,
            "is_in_collection": False,  # À implémenter
            "collections": [],
            "url": getattr(resource, 'url', ''),
            "thumbnail": getattr(resource, 'thumbnail', '')
        }
        result.append(resource_data)
    
    return result

@router.get("/resources/{resource_id}", response_model=dict)
def get_resource_by_id(
    resource_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Récupérer une ressource spécifique"""
    resource = db.query(Content).filter(Content.id == resource_id).first()
    if not resource:
        raise HTTPException(status_code=404, detail="Ressource non trouvée")
    
    # Vérifier si l'utilisateur a mis en favori
    is_favorite = db.query(UserFavorite).filter(
        UserFavorite.user_id == current_user.id,
        UserFavorite.content_id == resource.id
    ).first() is not None
    
    return {
        "id": resource.id,
        "title": resource.title,
        "description": resource.description,
        "type": resource.type,
        "subject": resource.subject,
        "level": resource.level,
        "tags": json.loads(resource.tags) if resource.tags else [],
        "author": resource.author,
        "created_at": resource.created_at.isoformat() if resource.created_at else "",
        "duration": getattr(resource, 'duration', None),
        "file_size": getattr(resource, 'file_size', None),
        "views": getattr(resource, 'views', 0),
        "rating": getattr(resource, 'rating', 0.0),
        "is_favorite": is_favorite,
        "url": getattr(resource, 'url', ''),
        "thumbnail": getattr(resource, 'thumbnail', '')
    }

# =====================================================
# ENDPOINTS POUR LES FAVORIS
# =====================================================

@router.post("/resources/{resource_id}/favorite")
def toggle_favorite(
    resource_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Ajouter/retirer une ressource des favoris"""
    # Vérifier que la ressource existe
    resource = db.query(Content).filter(Content.id == resource_id).first()
    if not resource:
        raise HTTPException(status_code=404, detail="Ressource non trouvée")
    
    # Vérifier si déjà en favori
    existing_favorite = db.query(UserFavorite).filter(
        UserFavorite.user_id == current_user.id,
        UserFavorite.content_id == resource_id
    ).first()
    
    if existing_favorite:
        # Retirer des favoris
        db.delete(existing_favorite)
        db.commit()
        return {"message": "Ressource retirée des favoris"}
    else:
        # Ajouter aux favoris
        new_favorite = UserFavorite(
            user_id=current_user.id,
            content_id=resource_id
        )
        db.add(new_favorite)
        db.commit()
        return {"message": "Ressource ajoutée aux favoris"}

@router.post("/resources/{resource_id}/collection")
def toggle_collection(
    resource_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Ajouter/retirer une ressource d'une collection par défaut"""
    # Vérifier que la ressource existe
    resource = db.query(Content).filter(Content.id == resource_id).first()
    if not resource:
        raise HTTPException(status_code=404, detail="Ressource non trouvée")
    
    # Récupérer ou créer une collection par défaut pour l'utilisateur
    default_collection = db.query(Collection).filter(
        Collection.user_id == current_user.id,
        Collection.name == "Mes Favoris"
    ).first()
    
    if not default_collection:
        # Créer une collection par défaut
        default_collection = Collection(
            name="Mes Favoris",
            description="Ressources que j'aime",
            user_id=current_user.id,
            is_public=False,
            color="bg-yellow-100"
        )
        db.add(default_collection)
        db.commit()
        db.refresh(default_collection)
    
    # Vérifier si la ressource est déjà dans la collection
    existing_item = db.query(CollectionItem).filter(
        CollectionItem.collection_id == default_collection.id,
        CollectionItem.content_id == resource_id
    ).first()
    
    if existing_item:
        # Retirer de la collection
        db.delete(existing_item)
        db.commit()
        return {"message": "Ressource retirée de la collection"}
    else:
        # Ajouter à la collection
        new_item = CollectionItem(
            collection_id=default_collection.id,
            content_id=resource_id
        )
        db.add(new_item)
        db.commit()
        return {"message": "Ressource ajoutée à la collection"}

# =====================================================
# ENDPOINTS POUR LES COLLECTIONS
# =====================================================

@router.get("/collections", response_model=List[dict])
def get_collections(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Récupérer les collections de l'utilisateur"""
    collections = db.query(Collection).filter(Collection.user_id == current_user.id).all()
    
    result = []
    for collection in collections:
        # Compter le nombre de ressources dans la collection
        resource_count = db.query(CollectionItem).filter(
            CollectionItem.collection_id == collection.id
        ).count()
        
        collection_data = {
            "id": collection.id,
            "name": collection.name,
            "description": collection.description,
            "resource_count": resource_count,
            "is_public": collection.is_public,
            "created_at": collection.created_at.isoformat() if collection.created_at else "",
            "color": collection.color
        }
        result.append(collection_data)
    
    return result

@router.post("/collections", response_model=dict)
def create_collection(
    collection_data: dict,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Créer une nouvelle collection"""
    db_collection = Collection(
        name=collection_data.get("name", ""),
        description=collection_data.get("description", ""),
        user_id=current_user.id,
        is_public=collection_data.get("is_public", False),
        color=collection_data.get("color", "bg-blue-100")
    )
    db.add(db_collection)
    db.commit()
    db.refresh(db_collection)
    
    return {
        "id": db_collection.id,
        "name": db_collection.name,
        "description": db_collection.description,
        "resource_count": 0,
        "is_public": db_collection.is_public,
        "created_at": db_collection.created_at.isoformat() if db_collection.created_at else "",
        "color": db_collection.color
    }

# =====================================================
# ENDPOINTS POUR LES MATIÈRES ET NIVEAUX
# =====================================================

@router.get("/subjects", response_model=List[dict])
def get_subjects(db: Session = Depends(get_db)):
    """Récupérer toutes les matières disponibles"""
    # Récupérer les matières uniques depuis les contenus
    subjects = db.query(Content.subject).distinct().all()
    
    result = []
    for subject in subjects:
        if subject[0]:  # Vérifier que le sujet n'est pas None
            # Compter le nombre de ressources pour cette matière
            resource_count = db.query(Content).filter(Content.subject == subject[0]).count()
            
            subject_data = {
                "id": len(result) + 1,  # ID temporaire
                "name": subject[0],
                "color": f"bg-blue-{500 - (len(result) * 100)}",
                "resource_count": resource_count
            }
            result.append(subject_data)
    
    return result

@router.get("/levels", response_model=List[dict])
def get_levels(db: Session = Depends(get_db)):
    """Récupérer tous les niveaux disponibles"""
    # Récupérer les niveaux uniques depuis les contenus
    levels = db.query(Content.level).distinct().all()
    
    result = []
    for level in levels:
        if level[0]:  # Vérifier que le niveau n'est pas None
            # Compter le nombre de ressources pour ce niveau
            resource_count = db.query(Content).filter(Content.level == level[0]).count()
            
            level_data = {
                "id": len(result) + 1,  # ID temporaire
                "name": level[0],
                "color": f"bg-green-{500 - (len(result) * 100)}",
                "resource_count": resource_count
            }
            result.append(level_data)
    
    return result

# =====================================================
# ENDPOINTS POUR LES RECOMMANDATIONS
# =====================================================

@router.get("/recommendations", response_model=List[dict])
def get_recommendations(
    limit: int = 10,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Récupérer les recommandations de ressources pour l'utilisateur"""
    # Logique simple de recommandation basée sur l'historique
    # À améliorer avec un algorithme plus sophistiqué
    
    # Récupérer les ressources les plus populaires
    popular_resources = db.query(Content).order_by(Content.views.desc()).limit(limit).all()
    
    result = []
    for resource in popular_resources:
        resource_data = {
            "id": resource.id,
            "title": resource.title,
            "description": resource.description,
            "type": resource.type,
            "subject": resource.subject,
            "level": resource.level,
            "rating": getattr(resource, 'rating', 0.0),
            "views": getattr(resource, 'views', 0)
        }
        result.append(resource_data)
    
    return result 